libraries/bot-ai-qna/src/main/java/com/microsoft/bot/ai/qna/QnAMakerRecognizer.java

```diff
@@ -0,0 +1,530 @@
+                    JSONObject data = new JSONObject() {
+                        {
+                            put("startIndex", 0);
+                            put("endIndex", activity.getText().length());
+                        }
+                    };
+                    instance.add(data);
+                    ObjectPath.setPathValue(recognizerResult, "entities.$instance.answer", instance);
+                    recognizerResult.setProperties("answers", answers);
+                } else {
+                    Map<String, IntentScore> intents = recognizerResult.getIntents();
+                    intents.put("None", new IntentScore() {
+                        {
+                            setScore(1.0f);
+                        }
+                    });
+                    recognizerResult.setIntents(intents);
+                }
+
+                this.trackRecognizerResult(dialogContext, "QnAMakerRecognizerResult",
+                        this.fillRecognizerResultTelemetryProperties(recognizerResult, telemetryProperties),
+                        telemetryMetrics);
+                return recognizerResult;
+            });
+        });
+    }
+
+    /**
+     * Gets an instance of {@link IQnAMakerClient}.
+     *
+     * @param dc The {@link DialogContext} used to access state.
+     * @return An instance of {@link IQnAMakerClient}.
+     */
+    protected CompletableFuture<IQnAMakerClient> getQnAMakerClient(DialogContext dc) {
+        IQnAMakerClient qnaClient = (IQnAMakerClient) dc.getContext().getTurnState().get();
+        if (qnaClient != null) {
+            // return mock client
+            return CompletableFuture.completedFuture(qnaClient);
+        }
+
+        OkHttpClient httpClient = (OkHttpClient) dc.getContext().getTurnState().get();
+        if (httpClient == null) {
+            httpClient = this.httpClient;
+        }
+
+        Pair endpointPair = this.endpointKey.tryGetValue(dc.getState());
+        Pair hostNamePair = this.hostName.tryGetValue(dc.getState());
+        Pair knowledgeBaseIdPair = this.knowledgeBaseId.tryGetValue(dc.getState());
+        Pair logPersonalInformationPair = this.logPersonalInformation.tryGetValue(dc.getState());
+
+        if (endpointPair.getLeft() == null) {
+            throw new IllegalStateException(String.format("Unable to get a value for %1$s from state. %2$s",
+                    "endpointKey", endpointPair.getRight()));
+        }
+
+        if (hostNamePair.getLeft() == null) {
+            throw new IllegalStateException(String.format("Unable to get a value for %1$s from state. %2$s", "hostName",
+                    hostNamePair.getRight()));
+        }
+
+        if (knowledgeBaseIdPair.getLeft() == null) {
+            throw new IllegalStateException(String.format("Unable to get a value for %1$s from state. %2$s",
+                    "knowledgeBaseId", knowledgeBaseIdPair.getRight()));
+        }
+
+        QnAMakerEndpoint endpoint = new QnAMakerEndpoint() {
```

Comment:
I have some doubts on this, do we need to add the default constructor to the class?